Add omit-return rule (#95)
airbnb/swift: README.md

  </details>

* <a id='omit-single-expression-return'></a>(<a href='#omit-single-expression-return'>link</a>) **Omit the `return` return keyword from single-expression functions accessors, and closures.** [![SwiftLint: omit-single-expression-return](https://img.shields.io/badge/SwiftLint-implicit-return-007A87.svg)](https://realm.github.io/SwiftLint/implicit_return.html)

<details>

```swift
// WRONG
["1", "2", "3"].compactMap { return Int($0) }

var size: CGSize {
  return CGSize(
    width: 100.0,
    height: 100.0)
}

> func buildInfoAlert(message: String) -> UIAlertController {

Reviewer: nit: the Swift API design guidelines specify that:
> Begin names of factory methods with “make”, e.g. x.makeIterator().

https://swift.org/documentation/api-design-guidelines/#strive-for-fluent-usage

So this would be `makeInfoAlert`

  </details>

> * <a id='omit-single-expression-return'></a>(<a href='#omit-single-expression-return'>link</a>) **Omit the `return` keyword when not required by the language.** [![SwiftFormat: redundantReturn](https://img.shields.io/badge/SwiftFormat-redundantReturn-7B0051.svg)](https://github.com/nicklockwood/SwiftFormat/blob/master/Rules.md#redundantReturn)

Reviewer 2: @calda should we update the link too, now that we've simplified the language?

Author: To just `#omit-return`?

Reviewer 2: That’s consistent with the `self` one, so yeah

> * <a id='omit-single-expression-return'></a>(<a href='#omit-single-expression-return'>link</a>) **Omit the `return` keyword from single-expression functions, accessors, and closures.** [![SwiftFormat: redundantReturn](https://img.shields.io/badge/SwiftFormat-redundantReturn-7B0051.svg)](https://github.com/nicklockwood/SwiftFormat/blob/master/Rules.md#redundantReturn)

Reviewer 3: Maybe:

```suggestion
* <a id='omit-single-expression-return'></a>(<a href='#omit-single-expression-return'>link</a>) **Omit the `return` keyword from single expressions when not required by the language.** [![SwiftFormat: redundantReturn](https://img.shields.io/badge/SwiftFormat-redundantReturn-7B0051.svg)](https://github.com/nicklockwood/SwiftFormat/blob/master/Rules.md#redundantReturn)
```

I don't think we need to be super specific about function, accessors and closures here

Author: @fdiaz This is a pedantic distinction, but in the [Swift grammar](https://docs.swift.org/swift-book/ReferenceManual/zzSummaryOfTheGrammar.html) the [`return`](https://docs.swift.org/swift-book/ReferenceManual/Statements.html#ID444) keyword isn't part of the [`expression`](https://docs.swift.org/swift-book/ReferenceManual/Expressions.html).

Here's the relevant grammar for a `function-body` (simplified to only show the relevant single-expression cases):
```
function-body → code-block

code-block → { statement }

statement → expression
statement → control-transfer-statement

control-transfer-statement → return-statement

return-statement → return expression 
```

In this grammar, `function-body` can either be `{ expression }` or `{ return expression }`. 

So within the Swift grammar, it's not strictly accurate to say that we can omit the `return` keyword _from an expression_. This grammar allows us to omit the `return` keyword from a single-expression function.

Reviewer 2: Could we simplify to this?

> **Omit the `return` keyword when not required by the language.**

Author: @bachand That sounds reasonable. There's existing precedent in the style guide for this sort of "required by the language" wording:

> Don't use self unless it's necessary for disambiguation or required by the language. ([link](https://github.com/airbnb/swift#omit-self))

Reviewer 2: I've been working on this style guide a long time @calda  😉
